### PR TITLE
tenant creation for everyone in public (with restrictions/gated by flag)

### PR DESCRIPTION
**Pull request by andreer**

### CODE REVIEW

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/Policy.java [hunk 1]
```diff
@@ -48,15 +48,10 @@
...
+    /** Access to create a tenant. */
     tenantCreate(Privilege.grant(Action.create)
                           .on(PathGroup.tenant)
-                          .in(SystemName.main, SystemName.cd, SystemName.dev)), // TODO SystemName.all()
-
-    /** Access to create a tenant in public */
-    tenantCreatePublic(Privilege.grant(Action.create)
-                                .on(PathGroup.tenant)
-                                .in(SystemName.PublicCd, SystemName.Public)),
+                          .in(SystemName.all())),
```
Comment (jonmv) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/Policy.java:54]: `(((＼（✘෴✘）／)))`

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java [hunk 2]
```diff
@@ -18,18 +18,17 @@
...
 
+    /** Base role which every user is part of. */
+    everyone(Policy.classifiedRead,
+            Policy.publicRead,
+            Policy.user,
+            Policy.tenantCreate),
+
     /** Deus ex machina. */
-    hostedOperator(Policy.operator),
+    hostedOperator(everyone, Policy.operator),
```
Comment (jonmv) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java:28]: The operator policy already provides all priviliges.

Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java:28]:
```suggestion
    hostedOperator(Policy.operator),
```

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java [hunk 3]
```diff
@@ -18,18 +18,17 @@
...
     /** Machina autem exspiravit. */
-    hostedSupporter(Policy.supporter,
-                    Policy.tenantCreatePublic),
-
-    /** Base role which every user is part of. */
-    everyone(Policy.classifiedRead,
-             Policy.publicRead,
-             Policy.user,
-             Policy.tenantCreate),
+    hostedSupporter(everyone, Policy.supporter),
```
Comment (jonmv) [inline controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java:31]:
```suggestion
    hostedSupporter(Policy.supporter),
```
Anyone who is authenticated is automatically a member of the `everyone` role as well, but I guess that could be considered an implementation detail. 
I do know someone disliked thet role inheritance, though, and wanted to kill it.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/security/Auth0Credentials.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    private final Set<Role> roles;
+
+    public Auth0Credentials(Principal user, Set<Role> roles) {
+        super(user);
+        this.roles = Collections.unmodifiableSet(roles);
+    }
+
+    @Override
+    public Principal user() { return super.user(); }
+
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/security/Auth0Credentials.java:27]:
```suggestion
```

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/security/CloudAccessControl.java [hunk 5]
```diff
@@ -48,14 +60,43 @@ public CloudTenant createTenant(TenantSpec tenantSpec, Credentials credentials,
...
+
     @Override
     public Tenant updateTenant(TenantSpec tenantSpec, Credentials credentials, List<Tenant> existing, List<Application> applications) {
         throw new UnsupportedOperationException("Update is not supported here, as it would entail changing the tenant name.");
     }
 
     @Override
     public void deleteTenant(TenantName tenant, Credentials credentials) {
-        // Probably terminate customer subscription?
+        // TODO: allow only if 0 resources, 0 balance
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/security/CloudAccessControl.java:99]: Yes!

**Review by jonmv — APPROVED** [review]
LGTM :) Do consider suggestions.

**Comment by jonmv** [comment]
Although test failure: 
```
[ERROR] Failures: 
[ERROR]   UserApiOnPremTest.userMetadataOnPremTest:57 user-without-applications.json (with ignored fields) expected:<[\Q{"isPublic":\E[^,:\[\]{}]*\Q,"isCd":\E[^,:\[\]{}]*\Q,"enable-public-signup-flow":true,"user":{"name":"Joe Developer","email":"dev@domail","nickname":"dev"},"tenants":{},"operator":["hostedOperator","hostedSupporter","hostedAccountant"]}\E]> but was:<[{"isPublic":false,"isCd":false,"user":{"name":"Joe Developer","email":"dev@domail","nickname":"dev"},"tenants":{},"operator":["hostedOperator","hostedSupporter","hostedAccountant"]}]>
```

**Comment by smorgrav** [comment]
What @jonmv said :)

**Comment by smorgrav** [comment]
The two visuals that fails here (one of the nodes view and the user workflow) are usual suspects when it comes to instability. Just revert them? 

The visual test experience is becoming a real pain point. With more tests, changing environments (no more localcd & ?) and more developers - this makes the entry-point for console participation very high. 
 
I think we should invest in these tests to smooth out the test experience. We have a related task that we could piggy back on for this https://jira.vzbuilders.com/browse/VESPA-17669.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/Policy.java (hunk 1 above)
Comment (andreer) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/Policy.java:54]: > I must not fear. Fear is the mind-killer. Fear is the little-death that brings total obliteration. I will face my fear. I will permit it to pass over me and through me. And when it has gone past I will turn the inner eye to see its path. Where the fear has gone there will be nothing. Only I will remain.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java (hunk 3 above)
Comment (andreer) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java:31]: True, but that does not apply when the hostedSupporter role is used in unit tests. I'll look at the affected tests and consider changing it there instead.

**Pull request merged**